app/uvector-vs-tba/Main.hs

```diff
@@ -0,0 +1,67 @@
+{-# LANGUAGE BangPatterns #-}
+
+-- | Microbenchmark to measure the difference in 'TypedByteArray' and 'Data.Vector.Unboxed.Vector' indexing performance.
+-- The later is a slice into a 'ByteArray' which means that every time you retrieve and element, an offset
+-- needs to be added to your index.
+--
+-- To reproduce:
+--
+-- @
+-- stack run uvector-vs-tba -- --output uvector-vs-tba.html
+-- @
+--
+-- You can pass a greater @--time-limit@ to increase the number of iterations.
+module Main where
+
+import Control.Monad.ST (runST)
+import Criterion.Main (Benchmark, bench, bgroup, defaultMain, nf)
+import Data.Foldable (for_)
+import Text.Printf (printf)
+
+import qualified Data.Vector.Unboxed as UVector
+import qualified Data.Vector.Unboxed.Mutable as UMVector
+
+import qualified Data.TypedByteArray as TBA
+
+main :: IO ()
+main = defaultMain
+  [ bgroup "tba" $ mkReadBenchs readTba genTba [7, 8]
+  , bgroup "uvector" $ mkReadBenchs readUVector genUVector [7, 8]
+  ]
+
+mkReadBenchs :: (Int -> a -> Int) -> (Int -> a) -> [Int] -> [Benchmark]
+mkReadBenchs readPat gen powers =
+  [ bench (printf "%d reads" n) $ nf (readPat n) $ gen n
+  | n <- map (10^) powers
+  ]
+
+readTba :: Int -> TBA.TypedByteArray Int -> Int
+readTba !n !arr = go 0
+  where
+    go !i
+      | i >= n    = 42
+      | otherwise = go $ TBA.unsafeIndex arr i
+
+readUVector :: Int -> UVector.Vector Int -> Int
+readUVector !n !arr = go 0
+  where
+    go !i
+      | i >= n    = 42
+      | otherwise = go $ UVector.unsafeIndex arr i
+
+-- NOTE: We should probably measure pseudo-random access time as well, e.g. by shuffing the generated arrays.
+
+genTba :: Int -> TBA.TypedByteArray Int
+genTba n = runST $ do
+  mutArr <- TBA.newTypedByteArray n
+  for_ [0 .. n-1] $ \i -> TBA.writeTypedByteArray mutArr i $ i + 1
+  TBA.unsafeFreezeTypedByteArray mutArr
+
+-- | Generate an unboxed vector @v@ such that @v[i] == i + 1@.
+-- This function also makes sure that the @offset@ property of the generated vector
+-- is not @0@ in order to avoid GHC optimizing that out.
```

Comment:
This one is a bit tricky since the optimizer was smart enough to remove the addition in some of the benchmarks I wrote. In the current implementation, it isn't. However, I left this bit so that future changes wouldn't accidentally change benchmark results.

Alternatively we could leave a warning here and remove the `slice` call.

Reply:
Should I remove the additional code then?

Reply:
Maybe with an additional `NOINLINE`, which hopefully prevents GHC from peeking into the definition.